src/common/KokkosKernels_HashmapAccumulator.hpp

```diff
@@ -492,44 +572,68 @@ struct HashmapAccumulator
       //hash_nexts is updated second time as below.
       //but this is okay for spgemm,
       //because no two keys will be inserted into hashmap at the same time, as rows have unique columns.
+      
       hash_nexts[my_write_index] = hash_begins[hash];
```

Comment:
@brian-kelley: Why is this here? Is this masking a cache consistency issue? When we are in this insertion routine, we are _supposed_ to be running from a single kokkos thread but in parallel from multiple vector lanes, right? So we can think of each of the kokkos thread partition of `keys`, `values`, hash_begins` or `hash_nexts` as shared among vector lanes but not accessed concurrently among kokkos threads, right? Is the loop above reading `keys` and `values` supposed to be protected from reading a stale/uninitialized element of `keys` or `values` through `hash_begins` and `hash_nexts`? Neither `keys`, `values`, `hash_begins` or `hash_nexts` are read from atomically. So, what if the write has not landed? Or, what if the write to `hash_nexts` has landed but the writes to `values` and `keys` have not landed? Does `key_type`, `value_type`, and `size_type` guarantee atomic reads and where are those types enforced?

Reply:
If a write to `keys` or `values` doesn't land before it's read, what guarantees that we do not get into the block starting at https://github.com/kokkos/kokkos-kernels/pull/731/files#diff-31dd0387f00dc52299083b47296b28b0R561? And, couldn't `values` at https://github.com/kokkos/kokkos-kernels/pull/731/files#diff-31dd0387f00dc52299083b47296b28b0R543 become corrupted in this scenario?

Reply:
@e10harvey That line is to make sure there is no break in the linked list observed by vector lane A, if vector lane B is in between lines 583 and 584 (in your branch). In that situation, lane B has replaced ``hash_begins[hash]`` by atomic exchange, but it hasn't yet made the link from ``my_write_index`` to ``hashbeginning``. Without line 576, at the instant that B is between 583 and 584, the list only contain's B's new node ``my_write_index`` (it's the hash begin, but its next is -1). Because of the assumption that we're not inserting the same key on multiple lanes at the same time, the invariant that matters is that walking the list gives you all keys that were previously inserted.

Reply:
If we didn't have this assumption, this insertion code wouldn't be correct at all. It would be necessary to insert at the end instead of the beginning, and it needs an atomic compare-and-swap, not an exchange.

Reply:
@brian-kelley I'm still trying to understand. Why are we not performing: `Kokkos::atmoic_store(hash_nexts+my_write_index, Kokkos::atomic_exchange(hash_begins+hash, my_write_index))` to ensure the linked list is not broken?

Reply:
@e10harvey That's not equivalent to doing atomic_exchange. Say the list was
```
begin -> 1
```
and vector lane A is inserting 2, and B is inserting 3 at the same time. After the first line is executed, the links are:
```
begin -> 1
2 -> 1
3 -> 1
```
then after the 2nd line, depending on the order that the 2nd atomic transaction happens, it could be either:
```
begin -> 2 -> 1
3 -> 1
```
or
```
begin -> 3 -> 1
2 -> 1
```
so one of the insertions fails.

In the same scenario, here's what the actual code does. After line 576:
```
begin -> 1
2 -> 1
3 ->1
```
Then one of them does the atomic exchange (without loss of generality, say it's A):
```
begin -> 2 -> 1, and hashbeginning is 1 on lane A
3 -> 1
```
Now, either B does the atomic exchange next, and then A and B both update nexts:
```
begin -> 3 -> 1. Hashbeginning is 1 on lane A, and 2 on lane B
2 -> 1
```
```
begin -> 3 -> 2 -> 1
```
OR, A updates its next first (doesn't change anything, since 2's next was already 1), and then B does the exchange followed by the next update:
```
begin -> 3 -> 1, and hashbeginning is 2 on lane B
2 -> 1
```
```
begin -> 3 -> 2 -> 1
```

So in any case, at all points in time, walking the list from begin will give you at least the keys that were in the list before (1). There are other ways the 3 lines could interleave, but I convinced myself that they all end up with 2 and 3 inserted, while always keeping 1 in the list. No matter what happens, the original list (starting at 1) is either reachable directly from begin, or from one of the new nodes. A new node can't replace begin until its next reaches the original list.

Reply:
I talked with Christian about what ``atomic_store`` does, and it's basically a volatile write (a normal write, but the compiler can't optimize away loads and stores involving that address across the atomic). There is no execution barrier before or after, so different threads/vector lanes can interleave like I described above

Reply:
Thanks Brian. What I am trying to ask is whether we should be performing:
```C
Kokkos::atomic_store(hash_nexts+my_write_index, Kokkos::atomic_exchange(hash_begins+hash, my_write_index));
```

In other words, is it OK for line 576 and 583 not to be atomic? I'm having trouble understanding but I *think* there could be a data race when updating hash_nexts here. I'll try to come up with a data race scenario.

Reply:
Lines 576 and 584 can't data race because they are both writing to ``hash_nexts[my_write_index]``, and ``my_write_index`` is guaranteed to be unique across all vector lanes and iterations, because it's counted up atomically.